gh-119258: Eliminate Type Guards in Tier 2 Optimizer with Watcher

saulshanabrook opened the pull request:
This PR replaces https://github.com/python/cpython/pull/119259 to more accurately remove `_GUARD_TYPE_VERSION`.

With help and guidance from @brandtbucher, I added a global type version watcher to the interpreter, so that we can look at types by their versions.

Then, whenever we see a `_GUARD_TYPE_VERSION` we can add a watcher to that type, and if it isn't changed, the next time we see that type version we can omit the guard.

We should see if the performance is improved compared to the other PR, or at least see if it removes more guard type versions.

<!-- gh-issue-number: gh-119258 -->
* Issue: gh-119258
<!-- /gh-issue-number -->

saulshanabrook committed:
Add type version and offset to optimizer header

saulshanabrook committed:
Start adding type version

saulshanabrook committed:
Add type version functions and struct initialization

saulshanabrook committed:
fix typo

saulshanabrook committed:
Add failing test case

dpdani committed:
breakpoints and prints

saulshanabrook committed:
Add function definitions

dpdani committed:
wooohoooo

dpdani committed:
unprint

saulshanabrook committed:
Remove generated printf

saulshanabrook committed:
Remove test breakpoints

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

saulshanabrook committed:
Revert "📜🤖 Added by blurb_it."

4060ab8e64488cf236dbbf8c4230501e8767443c

saulshanabrook committed:
Use global type watcher to invalidate type versions

saulshanabrook committed:
Revert formatting changes to tests

saulshanabrook committed:
Add additional test

saulshanabrook committed:
Remove printf and make slightly more threadsafe

saulshanabrook committed:
fix typo

saulshanabrook committed:
Update type watcher ID in test bc we know have a default one at 0

saulshanabrook committed:
Fix panic and add test for it

saulshanabrook committed:
add manual cast

saulshanabrook committed:
Properly fixed the frame creation with discussion with Ken Jin

saulshanabrook committed:
Change type_assign_specific_version_unsafe to use safe setting

We need to move type_assign_specific_version_unsafe to a different module
so we can use the private API to set the type version now, so it tracks
cleanup

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

brandtbucher committed:
Fix more places where tp_version_tag is being set manually

brandtbucher committed:
Merge branch 'main' into optimizer-type-version-watcher

brandtbucher committed:
_PyType_ClearCodeByVersion isn't necessary

Fidget-Spinner commented on Misc/NEWS.d/next/Core and Builtins/2024-05-23-20-17-37.gh-issue-119258.wZFIpt.rst:
```diff
@@ -0,0 +1,3 @@
+Eliminate type version guards in the tier two interpreter.
+
+Note that setting the ``tp_version_tag`` manually (which has never been supported) may result in crashes.
```
There's a test that sets it manually in `test/test_type_cache.py`. Could you decorate that with `without_optimizer` from `test.support` please?

brandtbucher replied:
I believe this test has already been updated to use the new API function, which updates the cache correctly.

saulshanabrook replied:
Yeah I believe that is correct, `type_assign_specific_version_unsafe` was moved to `_testinternalcapi` since it now uses an internal API and the tests were updated to use it.

Fidget-Spinner reviewed (commented):
lgtm mostly, just one comment. Thanks!

Fidget-Spinner commented on Python/optimizer_bytecodes.c:
```diff
@@ -113,6 +115,23 @@ dummy_func(void) {
+            // with the key, in which case we can't trust the version
+            if (type) {
+                sym_set_type_version(owner, type_version);
+                PyType_Watch(TYPE_WATCHER_ID, (PyObject *)type);
```
Just to be clear, this means at any point of time, we can only watch a single type in tier 2 right? I might be misunderstanding something here.

Fidget-Spinner replied:
Nevermind I misunderstood. The watcher is just a callback. We're just saying "use the same callback for all tier 2 watchers"

Fidget-Spinner committed:
Merge branch 'main' into optimizer-type-version-watcher

saulshanabrook committed:
Merge branch 'main' into optimizer-type-version-watcher

brandtbucher commented on Include/internal/pycore_optimizer.h:
```diff
@@ -33,6 +33,7 @@ struct _Py_UopsSymbol {
     int flags;  // 0 bits: Top; 2 or more bits: Bottom
     PyTypeObject *typ;  // Borrowed reference
     PyObject *const_val;  // Owned reference (!)
+    int32_t typ_version; // currently stores type version
```
Minor: can you change this to `type_version` everywhere?

```suggestion
    int32_t type_version;
```

brandtbucher replied:
Also, should this be an unsigned int, since that's what `tp_version_tag` is?

brandtbucher commented on Include/internal/pycore_typeobject.h:
```diff
@@ -108,6 +110,12 @@ struct types_state {
+    // tp_version_tag % TYPE_VERSION_CACHE_SIZE
+    // once was equal to the index in the table.
+    // They are cleared when the type object is deallocated.
+    PyTypeObject* type_version_cache[TYPE_VERSION_CACHE_SIZE];
```
```suggestion
    PyTypeObject *type_version_cache[TYPE_VERSION_CACHE_SIZE];
```

brandtbucher commented on Include/internal/pycore_typeobject.h:
```diff
@@ -61,6 +61,8 @@ typedef struct {
     PyObject *tp_weaklist;
 } static_builtin_state;
 
+#define TYPE_VERSION_CACHE_SIZE (1<<12)  /* Must be a power of 2 */
```
Not a big deal, but have we experimented with different values here? Might be useful to try out different values once this lands.

saulshanabrook replied:
Yeah I haven't... That would make sense. I also wonder about adding the possibility of adding some diagnostics metadata for cache misses?

brandtbucher commented on Lib/test/test_capi/test_opt.py:
```diff
@@ -1333,6 +1333,124 @@ def test_modified_local_is_seen_by_optimized_code(self):
+    def test_guard_type_version_removed_exiting(self):
+        """
+        Verify that the guard type version if we have an exiting function
+        """
```
```suggestion
    def test_guard_type_version_removed_inlined(self):
        """
        Verify that the guard type version if we have an inlined function
        """
```

brandtbucher commented on Lib/test/test_capi/test_opt.py:
```diff
@@ -1333,6 +1333,124 @@ def test_modified_local_is_seen_by_optimized_code(self):
+        guard_type_version_count = opnames.count("_GUARD_TYPE_VERSION")
+        self.assertEqual(guard_type_version_count, 1)
+
+    def test_guard_type_version_not_remove(self):
```
```suggestion
    def test_guard_type_version_not_removed(self):
```

brandtbucher commented on Lib/test/test_capi/test_opt.py:
```diff
@@ -1333,6 +1333,124 @@ def test_modified_local_is_seen_by_optimized_code(self):
+        self.assertIsNotNone(ex)
+        self.assertEqual(res, 181)
+        guard_type_version_count = opnames.count("_GUARD_TYPE_VERSION")
+        self.assertEqual(guard_type_version_count, 2)
```
I feel like this should be split into two tests:
- One that makes an escaping call where the `setattr` trick is now (maybe just a call to `sum` or something), and checking that both guards are there.
- One that performs the mutation partway through execution like this. However, maybe change `Foo.attr` to `2` so we know it keeps running, but can see the effect of the changed value? I'd also make `xxx` just another class `Bar`, since I'm worried that our function watcher branch might invalidate this or something weird since it's just a function.

saulshanabrook replied:
On the first bullet here, I think I may be missing something. If I had test like that (I tried with `eval` and `sum`), the second check is still optimized away.

Through briefly looking over the escaping logic, my understanding of it is:

1. When preparing the uops, we see if they may be escaping
2. When we have a `CHECK_VALID` uop, we remove it if we know we have no escaped uops before it. Otherwise, we keep it
3. Any op that escapes should at runtime set the `valid` field to be false
4. When the `CHECK_VALID` runs, it will bail out of execution the `valid` field is false.

So it would make sense that we still remove the second `_GUARD_TYPE_VERSION`, since the `CHECK_VALID` is in between them and would catch escaping functions? Am I understanding that right?

<details><summary>Details</summary>
<p>

Test:

```python
    def test_guard_type_version_not_removed_escaping(self):
        """
        Verify that the guard type version is not removed if have an escaping function
        """

        def thing(a):
            x = 0
            for i in range(100):
                x += a.attr
                # eval should be escaping and so should cause optimization to stop and preserve both type versions
                eval("None")
                x += a.attr
            return x

        class Foo:
            attr = 1
        res, ex = self._run_with_optimizer(thing, Foo())
        opnames = list(iter_opnames(ex))
        self.assertIsNotNone(ex)
        self.assertEqual(res, 200)
        guard_type_version_count = opnames.count("_GUARD_TYPE_VERSION")
        self.assertEqual(guard_type_version_count, 2)
```

Opcodes:

```
_START_EXECUTOR
_TIER2_RESUME_CHECK
_ITER_CHECK_RANGE
_GUARD_NOT_EXHAUSTED_RANGE
_ITER_NEXT_RANGE
_STORE_FAST_2
_LOAD_FAST_1
_LOAD_FAST_0
_GUARD_TYPE_VERSION
_GUARD_DORV_VALUES_INST_ATTR_FROM_DICT
_GUARD_KEYS_VERSION
_LOAD_ATTR_NONDESCRIPTOR_WITH_VALUES
_GUARD_BOTH_INT
_BINARY_OP_ADD_INT
_STORE_FAST_1
_CHECK_FUNCTION
_LOAD_CONST_INLINE_WITH_NULL
_LOAD_CONST_INLINE_BORROW
_SET_IP
_CALL_BUILTIN_FAST_WITH_KEYWORDS
_CHECK_PERIODIC
_CHECK_VALIDITY
_POP_TOP
_LOAD_FAST_1
_LOAD_FAST_0
_GUARD_DORV_VALUES_INST_ATTR_FROM_DICT
_GUARD_KEYS_VERSION
_LOAD_ATTR_NONDESCRIPTOR_WITH_VALUES
_GUARD_TOS_INT
_BINARY_OP_ADD_INT
_STORE_FAST_1
_JUMP_TO_TOP
_DEOPT
_EXIT_TRACE
_EXIT_TRACE
_ERROR_POP_N
_EXIT_TRACE
_DEOPT
_EXIT_TRACE
_ERROR_POP_N
_DEOPT
_DEOPT
_ERROR_POP_N
_DEOPT
_DEOPT
_EXIT_TRACE
_ERROR_POP_N
```


</p>
</details>

brandtbucher commented on Objects/typeobject.c:
```diff
@@ -1035,14 +1036,72 @@ type_mro_modified(PyTypeObject *type, PyObject *bases) {
+The Tier 2 interpreter requires looking up the type object by the type version, so it can install
+watchers to understand when they change.
+
+So we add a global cache from type version to weak references of type objects.
```
```suggestion
So we add a global cache from type version to borrowed references of type objects.
```

brandtbucher commented on Objects/typeobject.c:
```diff
@@ -1035,14 +1036,72 @@ type_mro_modified(PyTypeObject *type, PyObject *bases) {
+
+So we add a global cache from type version to weak references of type objects.
+
+This is similar to how the cache used with function objects and versions.
```
```suggestion
This is similar to func_version_cache.
```

brandtbucher commented on Python/optimizer_bytecodes.c:
```diff
@@ -113,6 +115,23 @@ dummy_func(void) {
+            // with the key, in which case we can't trust the version
+            if (type) {
+                sym_set_type_version(owner, type_version);
+                PyType_Watch(TYPE_WATCHER_ID, (PyObject *)type);
```
This needs to be checked for error.

Fidget-Spinner replied:
We don't check for the other watcher's erroring out either because they can never fail (we already reserved a slot for them). Mark's code for `PyDict_Watch` in this file doesn't check for error either.

saulshanabrook replied:
I added an assert here to verify this.

brandtbucher commented on Python/optimizer_bytecodes.c:
```diff
@@ -113,6 +115,23 @@ dummy_func(void) {
     }
 
+    op(_GUARD_TYPE_VERSION, (type_version/2, owner -- owner)) {
+        if (sym_matches_type_version(owner, type_version)) {
```
Staring at this a bit harder, it seems like there are a couple of different cases that we want to distinguish:

1. the symbol's version is 0, the expected version is 0
2. the symbol's version is nonzero, the expected version is 0
3. the symbol's version is 0, the expected version is nonzero
4. the symbol's version is nonzero, the expected version is nonzero, both match
5. the symbol's version is nonzero, the expected version is nonzero, both don't match

Cases 1 and 2 should be impossible, and we can `assert(type_version)` above this line.

Case 3 is what we're correctly handling in the `else` currently.

Case 4 is what we're correctly handling in the `if` currently.

Case 5 should result in narrowing `owner` to the "bottom" type, right? If so, this should probably happen in `sym_set_type_version`, which is called on line 127.

@Fidget-Spinner or @markshannon, can you confirm that my understanding is correct here?

saulshanabrook replied:
Thanks for thinking through all of these. For case 5, I am unclear how/when this could happen. This would only occur if `_CHECK_TYPE_VERSION` was ommitted twice, the first time with one type version, and then the second time with a different type version as an arg, but for the same symbolic value.

I am a little fuzzy on how these versions are generated. Are they captured based on the current type version of the type when it's generating the opcodes? If so, how could it be two different values?

If there is an example/test case here that would be helpful. Also is it possible to see in the test runner the additional args passed to each op, like the `type_version`?

Fidget-Spinner replied:
> Are they captured based on the current type version of the type when it's generating the opcodes? If so, how could it be two different values?

Yup the type version is the one at the time of specialization (bytecode rewriting). So if it just so happens that for example, a re-specialization is happening, the bytecode might have out of sync versions in their cache entries, leading to the scenario Brandt said in 5.

In general, we should hit the bottom if we're trying to set a type on something that is contradictory. E.g. if we're setting a type version on something that already has a type version. That is a contradiction. You can see an example test here https://github.com/python/cpython/blob/82bc177e9b116b4e816957a58105ebbd3ac21f81/Python/optimizer_symbols.c#L484

saulshanabrook replied:
That makes sense. I changed it so that it would set it to bottom if they are different versions (and not add the watcher b/c it must have already been added then).

After this change, I wonder if the logic would be clearer if it was all just in this block instead of in the get/set helper functions.

brandtbucher reviewed (commented):
Thanks for sticking with this! I have a few notes.

Also, can either @Fidget-Spinner or @markshannon confirm that the changes to `_Py_UOpsAbstractFrame` initialization are correct?

saulshanabrook committed:
Rename typ_version to type_version

suggested by @brandtbucher

saulshanabrook committed:
Change type of type_version to unsigned int

to match `typ_version_tag`

suggested by @brandtbucher

saulshanabrook committed:
C formatting fix

suggested by @brandtbucher

saulshanabrook committed:
Rephrase test names

suggested by @brandtbucher

saulshanabrook committed:
Rephrase comments

suggested by @brandtbucher

saulshanabrook committed:
Make existing inline test clear

As per comment by @brandtbucher:

> One that performs the mutation partway through execution like this. However, maybe change Foo.attr to 2 so we know it keeps running, but can see the effect of the changed value? I'd also make xxx just another class Bar, since I'm worried that our function watcher branch might invalidate this or something weird since it's just a function.

saulshanabrook committed:
Add a test for escaping behavior, but allow it to fail for now

suggested by @brandtbucher

saulshanabrook committed:
Verify that expected type version is non zero

suggested by @brandtbucher

saulshanabrook committed:
Fix typo

saulshanabrook committed:
Assert pytype watch never errors

saulshanabrook committed:
Fix assertion

saulshanabrook committed:
Remove unnecessary assert

saulshanabrook committed:
Handle if type versions are different

https://github.com/python/cpython/pull/119365#discussion_r1630593619
